runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkStateInternals.java

```diff
@@ -622,4 +627,82 @@ public Boolean read() {
       };
     }
   }
+
+  private final class SparkOrderedListState<T> extends AbstractState<List<TimestampedValue<T>>>
+      implements OrderedListState<T> {
+
+    private SparkOrderedListState(StateNamespace namespace, String id, Coder<T> coder) {
+      super(namespace, id, ListCoder.of(TimestampedValue.TimestampedValueCoder.of(coder)));
+    }
+
+    private SortedMap<Instant, TimestampedValue<T>> readAsMap() {
+      final List<TimestampedValue<T>> listValues =
```

Comment:
The reason for each additional kind of state is to efficiently offer a novel form of a state access. The state access here as the same performance characteristics as `ValueState`. It is actually better for the runner to reject a pipeline than to run it with performance characteristics that don't match the expected performance contract.

Is there some underlying mechanism in Spark that could implement OrderedListState efficiently and scalably?

Reply:
I agree with your point. Let me share my thoughts on why I chose this implementation.

I've noticed that ListState/OrderedListState is mostly used in situations where writes happen much more frequently than reads. That's why I went with ArrayList instead of SortedMap - it's simply better at handling these frequent writes.

When it comes to reading data, it usually happens in just a couple of scenarios - either during OnTimer execution or when the list hits a certain size. So even if the read performance takes a small hit, it's not really going to affect the overall performance much.

It's also worth mentioning that FlinkOrderedListState uses the same approach, which gives me confidence in this design choice.

That's why I think the current implementation makes more sense for real-world usage patterns.

Reply:
I see. If Flink is implemented then it is OK with me to follow that precedent. My point was that this does not actually add capability that is more than `ValueState` provides. It is just a minor API wrapper adjustment - still useful but not the main purpose.

So we can merge with this design. But if you think about following up, here is how we would really like this to behave:

 - `add` should call some native Spark API that writes the element without reading the list
 - `readRange` should only read the requested range, ideally seeking in near-constant time (aka without a scan or sort)
 - `clearRange` should also seek in near-constant time
 - `isEmpty` should not read the list

Reply:
@kennknowles 

Hello. I've been reviewing the Spark documentation over the weekend.

Unfortunately, as far as I can tell, Spark doesn't have the native state API that we're looking for. Moreover, since the current SparkStateInternals is a class implemented independently of the Spark API, it's also difficult to apply RDD-based APIs.

The performance requirements you mentioned seem to be well satisfied by `WindmillOrderedList`. Given that the current implementation cannot guarantee the same performance, I completely understand if the PR doesn't get merged.

However, regardless of whether this PR gets merged or not, I will continue researching to meet the requirements you've outlined. I plan to explore better implementation approaches, taking inspiration from concepts like pendingAdds and pendingDeletes as used in `WindmillOrderedList`.

Belated Merry Christmas.